### DISCUSSION TITLE
CrashLoopBackOff, kube-green restarts and CPU/memory limits doubts

### DISCUSSION THREAD

**Discussion by SalvadorCali** [mia-platform/community#82] [Q&A] [answered]
Hi everyone!
We use kube-green in dev and test environments to turn off our projects during the night. 
The 7th of october we revised CPU and memory limits for some of our services and deployed their projects. 
Starting from the 17th of october we noticed that many services go everyday in CrashLoopBackOff state after the morning restart of kube-green and we thought that this problem could be caused by the limits revision.

![image](https://user-images.githubusercontent.com/37811007/200784512-0816db89-733c-44e9-b0d4-475b3191c709.png)

We have many doubts regarding this topic:
- if this problem was caused by the limits revision, why doesn't it appear since the 7th of october?
- we noticed that one service ended up in the CrashLoopBackOff state although its limits were the same;
- we noticed that one service ended up in the CrashLoopBackOff state although its limits were higher than the previous ones.

Do you have any advice to solve this problem? Is it just a matter of limits or there is something else?

**Comment by davidebianchi** [comment, marked as answer]
Hi @SalvadorCali! Yes, this could be an issue with the request and limits set for your services.

How many namespaces do you sleep with kube-green? A simple suggestion could be to spread the time of wake up of a few minutes. In this way, there will be fewer services which start at the same time.

I try to answer to your questions:

> if this problem was caused by the limits revision, why doesn't it appear since the 7th of october?

This depends by the nodes available resources. Maybe the limits are ok, but the request is too low and resources in the node are just enough for the request of all the services. In this way, all the services cannot use resources given by limits because not present in the node. So, how are the nodes capacity of the cluster? After the 7th october, do you have add some service to the cluster?

> we noticed that one service ended up in the CrashLoopBackOff state although its limits were the same;

Also in this case, the reason may be the one described in the answer above

> we noticed that one service ended up in the CrashLoopBackOff state although its limits were higher than the previous ones.

Also in this case, the reason may be the one described in the answer above

Remember that the only allocated resource in the node (and cluster) are the resources set as request. Limits are an upper bound, if available resource are present in the node.
So, I suggest you to review the request of your services, and if you can't raise it up change the liveness probe to increase the time of start up.

**Reply by SalvadorCali** [reply]
> How many namespaces do you sleep with kube-green? A simple suggestion could be to spread the time of wake up of a few minutes. In this way, there will be fewer services which start at the same time.

Around 50. We just tried to spread the wake up time of some namespaces that restart at 7.
Do you think that automatically defer the restart of the namespaces with the same wakeUpAt could be an useful improvement for kube-green?

Anyway, thank you for your answer. We'll follow your advice and investigate more in the following days.

**Reply by davidebianchi** [reply]
For this, it should be open an issue on kube-green repo. Anyway, it could be but it is not so simple to make it automatically.
